### ISSUE TITLE
HA Masters do not provide Failover

### ISSUE THREAD

**Issue by dvohra** [openshift/openshift-ansible#2277]
HA Masters do not provide failover.

Cluster:
2 HA masters (not schedulable)
1 worker node
total 3 nodes
1 etcd
1 lb

Start cluster and access cluster from a master node. Create a deployment and run some Pods, which run fine. Shutdown the master node from which the cluster is accessed and access cluster from the other master. A deployment does not get deployed and Pods do not get run.

**Comment by abutcher** [comment]
Could you be more specific with this test case? With a cluster layout as described, say I have an application comprised of 3 identical pods (one per node). If I shut down 1 master then I would expect that the 3 pods continue to run without change. I would then be able to access cluster resources such as to scale the application down to 2 pods and back up to 3 pods with one of the masters down.

**Comment by dvohra** [comment]
_Could you be more specific with this test case?_

1. Create a cluster as described previously. The HA masters are listed in sequence in [masters] in /etc/ansible/hosts.

2. Access cluster from master listed first in [masters]
3. Create a deployment 
```
oc new-app openshift/deployment-example:v1 
oc describe dc/deployment-example
oc get pods
```

4. Shutdown the master from which cluster is accessed.
5. Login to the other master. The api server is still running as it is running on the load balancer IP.
6. The Pods deployed on the first master do not get listed with `oc get pods`
7. Create a deployment on the 2nd master with same commands; a new deployment gets created with errors and so does the Pod have Error and is not Running.

**Comment by abutcher** [comment]
Both masters will be communicating with the same etcd cluster so the original deployment won't need to be recreated and will be maintained by the controllers running on the second master (if the first master were to be shutdown).

**Comment by dvohra** [comment]
What you indicated is the expected result, but did not get the same result. Instead the deployment was not listed and a new deployment from the same or different Docker image could not be created. Could it be verified?

**Comment by abutcher** [comment]
Right, I think this is due to a misconfiguration that occurred during previous installation failures. If the original deployment and pods no longer exist when one master is shut down then each master would have to be talking to a different data store. I will verify your test case locally as soon as I can.

**Comment by dvohra** [comment]
What is the "misconfiguration that occurred during previous installation failures."?
The first deployment from the master listed first in [masters] runs fine. Only, after the first master is shutdown and cluster accessed from the 2nd master does a deployment not get listed or created.

**Comment by abutcher** [comment]
Does each master have the same `etcdClientInfo.urls` set in `/etc/origin/master/master-config.yaml`?

**Comment by dvohra** [comment]
Have not modified any default settings. If some settings are required to be modified for HA it is not mentioned in the OpenShift documentation.
https://docs.openshift.org/latest/install_config/install/advanced_install.html

**Comment by abutcher** [comment]
The multiple masters section has some required options for HA.

https://docs.openshift.org/latest/install_config/install/advanced_install.html#multiple-masters

```
# Native high availbility cluster method with optional load balancer.
# If no lb group is defined installer assumes that a load balancer has
# been preconfigured. For installation the value of
# openshift_master_cluster_hostname must resolve to the load balancer
# or to one or all of the masters defined in the inventory if no load
# balancer is present.
openshift_master_cluster_method=native
openshift_master_cluster_hostname=openshift-cluster.example.com
openshift_master_cluster_public_hostname=openshift-cluster.example.com
```

**Comment by dvohra** [comment]
The HA requirements have been set to the load balancer DNS.
```
openshift_master_cluster_method=native
openshift_master_cluster_hostname=openshift-cluster.example.com
openshift_master_cluster_public_hostname=openshift-cluster.example.com
```

**Comment by dvohra** [comment]
Is a Failover configured, because by itself a load balancer HAProxy does not provide a failover and failover is not even mentioned in the HA documentation link. Isn't a HAProxy Router required to replace the standard router?

**Comment by abutcher** [comment]
Sorry for the delay in responding!

The `[lb]` group configured by openshift-ansible is a single point of failure. Any other load balancing solution should be used in a production environment (elb, f5, haproxy+keepalived, ...) to balance the master API port (default 8443). We call attention to this in the docs but it could be made much clearer that the haproxy nodes are not configured active-passive.

The master API can be load balanced (+ health checks). Only one master (controllers service) will maintain a lease with etcd and if the master currently holding the lease is stopped then another master will take the lease. There is no need to configure failover for the masters unless it is desired for some reason.

The router can also be load balanced (+ health checks). Since the router uses host ports, it's best to set aside a few nodes with an "infrastructure" label matching the selector of the router deployment and load balance traffic the router port(s) on those nodes.

**Comment by dvohra** [comment]
Essentially, the HA Masters is to distribute the load across the master nodes as the api server is exposed  on the Load balancer IP.

Also, have not tested accessing the cluster from the master listed 2nd in [masters]. Could the cluster be accessed from any master? Or just the master listed first in [masters]?

**Comment by abutcher** [comment]
The cluster can be accessed from any master.

**Comment by dvohra** [comment]
Thanks. The HA Master is designed to provide load balancing on the HA Masters cluster, and is not designed for failure of a master node or failover. If one master is shutdown the cluster becomes inaccessible according to testing by shutting down one master.

**Comment by detiber** [comment]
@dvohra there is a currently known issue where the master is not removed from the service load balancer under certain conditions. I believe the fix for this is available in the latest 1.3 alpha builds.

**Comment by dvohra** [comment]
Jason, thanks for the update.
